docs/design-docs/benchmarks.md

```diff
@@ -0,0 +1,262 @@
+3. `BenchmarkRegistry` provides a way of benchmark registration, maintains all the registered benchmarks, and supports benchmark launching by `BenchmarkContext`.
+4. `BenchmarkContext` provides the context to launch one benchmark, including name, parameters, platform(CPU, GPU, etc.), and framework(Pytorch, TF, ONNX, etc.).
+5. `BenchmarkResult` defines the structured results for each benchmark in json format, including name, return_code, start_time, end_time, raw_data, summarized metrics, etc.
+
+The `Executor` on the uppermost layer is the entrance for all the benchmarks. It launches the benchmark by `BenchmarkRegistry` and fetch `BenchmarkResult`.
+
+## Detailed Component Design
+
+This chapter will describe the design details of all the components in `benchmarks` package.
+
+### E2E Model Benchmarks
+
+The E2E model benchmarks have 4-layer inheritance relationship.
+
+#### Training 
+
+The general process of model training is:
+
+init_distributed_setting -> generate_dataset -> init_dataloader -> create_model -> create_optimizer -> train
+
+These functions will be executed according to the order in the following figure. The functions that exist in derived class but not in base class are abstract functions.
+
+![Training process and the responsibility for function implementation of every layer](../assets/model-training-process.png)
+
+#### Inference
+
+The general process of the model inference is:
+
+Init_distributed_setting -> generate_dataset -> init_dataloader -> create_model -> inference
+
+Compared with training, it just gets rid of create_optimizer operation.
+
+![Inference process and the responsibility for function implementation of every layer](../assets/model-inference-process.png)
+
+### Micro Benchmarks
+
+The micro-benchmarks have 3-layer Inheritance Relationship. There have two base class for micro-benchmark: 
+One is for pure-python benchmarks named `MicroBenchmark`.
+Another is for benchmarks depending on third-party executable program named `MicroBenchmarkWithInvoke`.
+
+![micro-benchmarks benchmarking process for MicroBenchmark and the responsibility for function implementation of every layer](../assets/micro-benchmark-process-python.png)
+
+![micro-benchmarks benchmarking process for MicroBenchmarkWithInvoke and the responsibility for function implementation of every layer](../assets/micro-benchmark-process-native.png)
+
+### Docker Benchmarks
+
+The Docker benchmarks have 3-layer Inheritance Relationship. The Docker benchmarks need docker env ready.
+
+![docker-benchmarks benchmarking process and the responsibility for function implementation of every layer](../assets/docker-benchmark-process.png)
+
+### BenchmarkRegistry
+
+`BenchmarkRegistry` is designed to
+1.	Provide a way of benchmark registration.
+2.	Avoid modifying existing code when adding new benchmarks.
+3.	Reduce the redundant code for benchmarks with only different configurations, such as models with different scale of parameters.
```

Comment:
Reduce the redundant code for benchmarks with different configurations

Reply:
Done